### PR TITLE
Move request open timestamp to the top of the stack

### LINKED ISSUE #479: Reported request latency telemetry doesn't cover all the overhead of proxy
In https://github.com/runconduit/conduit/issues/437#issuecomment-368691604, @seanmonstar wrote:
> I don't have any leads yet on to why the latency is increasing, but I think I do know why Conduit isn't reporting it: the http Event::StreamRequestOpen is triggered just before we serialize the request, and so any time since having accepted the request originally and waiting on DNS, connecting, and waiting in queue is not calculated in the latency.

In https://github.com/runconduit/conduit/issues/437#issuecomment-369022602, @seanmonstar wrote:
> So, we do collect telemetry on both client and server socket opens and closes, but it looks like those events don't get forward into prometheus, and so cannot be shown in the conduit dashboard.

One summary of this is that we're measuring mostly the service's latency but we're not measuring all the additional latency that Conduit adds. Thus users looking at the Conduit Web UI and/or `conduit stat` output may be getting inaccurate, overly-optimistic numbers.

This matches the behavior initially reported in #437 where @bourquep that he could measure the latency increasing but Conduit's UI was reporting the latency not increasing.

Assigning to @olix0r so he can set the prioritization for the 0.3.1 or 0.4 release and resource it.

/cc @adleong

### PR DESCRIPTION
**Pull request by hawkw**
Currently, the request open timestamp, which is used for calculating latency, is captured in the `sensor::http::Http` middleware. However, the sensor middleware is placed fairly low in the stack, below some of the proxy's components that can add measurable latency (e.g. the router).

This PR moves the request_open timestamp out of the `Http` middleware and into a new `TimestampRequestOpen` middleware, which is installed at the top of the stack (before the router). The `TimestampRequestOpen` middleware adds the timestamp as a request extension, so that it can later be consumed by the `Http` sensor to generate the request stats.

By moving the timestamping to the top of the stack, the timestamp should more accurately cover the overhead of the proxy, but a majority of the telemetry work can still be done where it was previously.

I'd like to have included unit tests for this change, but since the expected improvement is in the accuracy of latency measurements, there's no easy way to test this programmatically. 

Fixes #479.

### CODE REVIEW

File: proxy/src/lib.rs [hunk 1]
```diff
@@ -366,7 +366,12 @@ where
                     http::StatusCode::INTERNAL_SERVER_ERROR
                 }
             }
-        })
+        });
+
+        // Install the request open timestamp module at the very top
+        // of the stack, in order to take the timestamp as close as
+        // possible to the beginning of the request's lifetime.
+        telemetry::sensor::http::TimestampRequestOpen::new(map_err)
```
Comment (olix0r) [inline proxy/src/lib.rs]: let's put this around `router` instead of around `map_err`, so that we still uniformly handle errors. Basically, i think that MapErr should wrap _everything_ unless there's a good reason not to.

Comment (hawkw) [reply proxy/src/lib.rs]: Okay, sounds good.

**Comment by olix0r** [comment]
to test this i might do something like the following:

- off of master -- add a significant delay into the call stack (probably within the router), confirm that the proxy doesn't report that added latency.
- off of this branch -- add the same delay and confirm that the proxy _does_ report the added latency.

i'd do this for at least http/1 (i tend to use curl/twistd, but anything is fine) and http/2 requests (i tend to use nghttp/nghttpd).

File: proxy/src/telemetry/sensor/http.rs [hunk 2]
```diff
@@ -15,6 +15,25 @@ use telemetry::event::{self, Event};
 
 const GRPC_STATUS: &str = "grpc-status";
 
+/// A `RequestOpen` timestamp.
+///
+/// This is added to a request's `Extensions` by the `TimestampRequestOpen`
+/// middleware. It's a newtype in order to distinguish it from other
```
Comment (briansmith) [inline proxy/src/telemetry/sensor/http.rs]: IMO "It's a newtype..." isn't useful.

File: proxy/src/telemetry/sensor/http.rs [hunk 3]
```diff
@@ -574,3 +595,48 @@ impl BodySensor for RequestBodyInner {
...
+    type Error = S::Error;
+    type Future = S::Future;
+
+    fn poll_ready(&mut self) -> Poll<(), Self::Error> {
+        self.inner.poll_ready()
+    }
+
+    fn call(&mut self, mut req: Self::Request) -> Self::Future {
+        let request_open = Instant::now();
+        req.extensions_mut().insert(RequestOpen(request_open));
```
Comment (briansmith) [inline proxy/src/telemetry/sensor/http.rs]: If the intent is that we really capture the earliest instant that we have the request, then why not just add this extension at the time we constructed `req`, instead of here? In particular, it's really unclear to me what happens between the time we constructed `req` and the time `call()` is called here, and it's unclear how we know that there are no non-trivial delays between those two things.

In other words, while it seems clean to implement this as middleware, I'm not convinced that implementing it as middleware most accurately captures the data.

File: proxy/src/telemetry/sensor/http.rs [hunk 4]
```diff
@@ -15,6 +15,25 @@ use telemetry::event::{self, Event};
 
 const GRPC_STATUS: &str = "grpc-status";
 
+/// A `RequestOpen` timestamp.
+///
+/// This is added to a request's `Extensions` by the `TimestampRequestOpen`
+/// middleware. It's a newtype in order to distinguish it from other
+/// `Instant`s that may be added as request extensions.
+#[derive(Copy, Clone, Debug)]
+pub struct RequestOpen(pub Instant);
```
Comment (briansmith) [inline proxy/src/telemetry/sensor/http.rs]: What does "Open" mean? Is there another term that would better capture this? E.g. "Parsed"?

What is this measuring and what *isn't* it measuring? I think it's important to document that this does not and cannot measure the time the request spend in the TCP receive window and any time spend parsing and/or otherwise pre-processing the request header. Simlarly, we should have some note about what our expectations are on the response end regarding unmeasured overhead. It would be also good to note any other overhead that isn't accounted for.

File: proxy/src/telemetry/sensor/http.rs (hunk 3 above)
Comment (hawkw) [reply proxy/src/telemetry/sensor/http.rs]: @briansmith I agree with your concerns; however, it appears that the request is constructed in a library (either in `h2` [here](https://github.com/carllerche/h2/blob/fabae35311362cf24fa9e2e08e27c3d64f6c67b1/src/server.rs#L493) or in `hyper` [here](https://github.com/hyperium/hyper/blob/dfdca25c004aeb583b931f57e19510b3096362e3/src/proto/h1/dispatch.rs#L355-L359), depending on the request protocol). The middleware is the earliest point that _Conduit_ has access to the request struct.

I didn't think that PRs against those libraries in order to add a Conduit-specific extension seemed likely to be accepted, and therefore, forking our library dependencies didn't seem worth what appears to be a fairly small gain in accuracy, as the service is called with the constructed request shortly thereafter. I _did_ consider using the [`serve_modified`](https://tower-rs.github.io/tower-h2/tower_h2/server/struct.Server.html#method.serve_modified) method in `tower-h2` to add the timestamp header to the request earlier in the pipeline, but there's no similar method in `hyper`, so there would be a difference in accuracy between the HTTP/1 and HTTP/2 timestamps. I felt that using a middleware, while less accurate overall, results in more consistent measurements between both protocols.

If you have any ideas I may have overlooked, I'd appreciate your suggestion, and I'd also like to hear what @olix0r thinks, as well.

Comment (briansmith) [reply proxy/src/telemetry/sensor/http.rs]: OK. Thinking in terms of middleware, then, wouldn't the ideal be for the layer that adds the request extension to be the later directly wrapping the hyper/h2 layers?

Comment (hawkw) [reply proxy/src/telemetry/sensor/http.rs]: I'm not sure if I entirely understand what you mean --- by placing this middleware at the top of the stack (modulo https://github.com/runconduit/conduit/pull/744#discussion_r180920549), it should be among the first layers to see the request object once its' been constructed.

Comment (briansmith) [reply proxy/src/telemetry/sensor/http.rs]: > I'm not sure if I entirely understand what you mean --- by placing this middleware at the top of the stack (modulo #744 (comment)), it should be among the first layers to see the request object once its' been constructed.

How could we ensure that it is definitely the first layer to see the request object?

Comment (hawkw) [reply proxy/src/telemetry/sensor/http.rs]: 1. We wrap the rest of the `NewService` stack with the `TimestampRequestOpen` `NewService`, so it's at the top of the stack:
https://github.com/runconduit/conduit/blob/be9be0ccab97680df90f814884634c5479f02f56/proxy/src/lib.rs#L374
2. that stack of `NewService`s is passed to `Server::new`, and goes in the constructed `Server`'s `new_service` field 
3. when the `Server` serves an HTTP/1 request, it calls `new_service` on that stack, constructing a new `Service` where the `TimestampRequestOpen` middleware is the outermost service (the first to see the request object). It's wrapped in an adapter that turns a `tower::Service` into a `hyper::Service`, but we can't put it before that because  the version of `hyper` we depend on uses the `hyper::Request` type rather than `http::Request`, and we can't add extensions to a `hyper::Request`:
https://github.com/runconduit/conduit/blob/be9be0ccab97680df90f814884634c5479f02f56/proxy/src/transparency/server.rs#L155-L169
4. When the server is constructed, it constructs a `tower_h2::Server` with the `stack`
https://github.com/runconduit/conduit/blob/be9be0ccab97680df90f814884634c5479f02f56/proxy/src/transparency/server.rs#L73-L81
which is used to serve HTTP/2 connections:
https://github.com/runconduit/conduit/blob/be9be0ccab97680df90f814884634c5479f02f56/proxy/src/transparency/server.rs#L178

---

In both cases, there are protocol-specific adapter middlewares around the rest of the middleware stack (`HttpBodyService` and `HyperServerService`),. In the case of the `HyperServerService` for HTTP/1, we can't add the timestamp before that service, for the reason discussed above. The H2  `HttpBodyService` just does a type conversion on the request body. We could place the middleware before that service, although I don't think it has a significant impact on latency, and this would complicate the implementation. What do you think?

Comment (olix0r) [reply proxy/src/telemetry/sensor/http.rs]: > How could we ensure that it is definitely the first layer to see the request object?

I don't think it should be the _first_ layer. (I think the error-handling/logging layer should be the first layer, so it can capture any errors emitted by middlewares like this); though it's true that the timestamping middleware can't fail on its own, so this is all semantics.

For now, I think the best we can do to _ensure_ that this is early in the stack is to document it well.

In the future, I hope that we can have better stacking primitives so these assurances can be made more explicitly (as in finagle stacks).

**Comment by briansmith** [comment]
It seems like at least @hawkw and @olix0r agree that the proposed implementation is OK. It at least improves the amount of the stack that we measure. I think that, in the future, we can improve the accounting further, and/or with greater confidence that we're measuring as much as we can, and I don't think we should consider the issue of accounting for all the overhead of the proxy that we can account for to be resolved. But, also, I think it is fine to merge this PR as-is.

File: proxy/src/telemetry/sensor/http.rs (hunk 3 above)
Comment (olix0r) [reply proxy/src/telemetry/sensor/http.rs]: i meant "error logging." I don't feel super strongly about this -- it makes some sense to just put this at the top of the stack. I don't think we have to spend too much time trying to build solid assurances that this is always at the top (yet) -- though I agree that we should at least begin to document the stack ordering.

File: proxy/src/telemetry/sensor/http.rs (hunk 4 above)
Comment (hawkw) [reply proxy/src/telemetry/sensor/http.rs]: The name "RequestOpen" was chosen because it was previously used here
https://github.com/runconduit/conduit/blob/e9b209829d8b5514b89850bb0b7a64a28b3348b1/proxy/src/telemetry/sensor/http.rs#L215

and elsewhere in the module. I'll improve the documentation.

File: proxy/src/telemetry/sensor/http.rs [hunk 5]
```diff
@@ -212,7 +231,9 @@ where
                 self.handle
                     .send(|| Event::StreamRequestOpen(Arc::clone(&ctx)));
 
-                let request_open = Instant::now();
+                let RequestOpen(request_open) = req.extensions_mut()
+                    .remove::<RequestOpen>()
+                    .expect("request open timestamp should always be inserted");
```
Comment (olix0r) [inline proxy/src/telemetry/sensor/http.rs]: Can we be more defensive than panicking?  Above, we are graceful to missing Server context, for instance... it seems better to fail silently (possibly logging) than to fail hard here.

Comment (hawkw) [reply proxy/src/telemetry/sensor/http.rs]: Will do.

File: proxy/src/telemetry/sensor/http.rs [hunk 6]
```diff
@@ -222,19 +222,25 @@ where
...
-            Some(ctx) => {
+        let metadata = (
+            req.extensions_mut().remove::<Arc<ctx::transport::Server>>(),
+            req.extensions_mut().remove::<RequestOpen>()
+        );
+        let (inner, body_inner) = match metadata {
+            (ctx @ None, request_open) | (ctx, request_open @ None) => {
+                warn!(
+                    "missing metadata for a request to {:?}; ctx={:?}; request_open={:?};",
+                    req.uri(), ctx, request_open
```
Comment (hawkw) [inline proxy/src/telemetry/sensor/http.rs]: Note that we log only the request URI here, because logging the entire request struct would require a `A: Debug` bound on the request's body type, which I didn't want to add.

File: proxy/src/telemetry/sensor/http.rs (hunk 5 above)
Comment (hawkw) [reply proxy/src/telemetry/sensor/http.rs]: @olix0r what do you think of 55ee99876a0dc77017e884aed87cccf9fe7a1a1b?

File: proxy/src/telemetry/sensor/http.rs [hunk 7]
```diff
@@ -203,17 +222,25 @@ where
...
     fn call(&mut self, mut req: Self::Request) -> Self::Future {
-        let (inner, body_inner) = match req.extensions_mut().remove::<Arc<ctx::transport::Server>>() {
-            None => (None, None),
-            Some(ctx) => {
+        let metadata = (
+            req.extensions_mut().remove::<Arc<ctx::transport::Server>>(),
+            req.extensions_mut().remove::<RequestOpen>()
+        );
+        let (inner, body_inner) = match metadata {
+            (ctx @ None, request_open) | (ctx, request_open @ None) => {
```
Comment (olix0r) [inline proxy/src/telemetry/sensor/http.rs]: alternatively, just match on the `(Some(), Some())` case first and fall back to `(ctx, request_open)`

Comment (hawkw) [reply proxy/src/telemetry/sensor/http.rs]: Yeah, that's much clearer, changed in 0b6014fcaae38aa8f2e53626f75ac0c72d20d458
